### ISSUE TITLE
RFC: vulnerability reachability

### ISSUE THREAD

**Issue by TG1999** [nexB/vulnerablecode#1313]
Vulnerability reachability is to check if vulnerable code is reachable or not.
This is important to help triage vulnerabilities.

Some of the things to consider:
- [ ] Collecting fix commits or patches to find vulnerable functions
- [ ] Finding the path in the affected package code graph to the vulnerable functions
- [ ] Finding the path in the codebase under analysis that may use the affected package vulnerable functions

**Comment by pombredanne** [comment]
A unfinished attempt to collect fix commits is in https://github.com/nexB/vulnerablecode/pull/1226 and I am closing it. We can reuse some of it as inspiration for this feature here.

**Comment by JafarAkhondali** [comment]
@pombredanne Sorry I was busy with my research, I've created a new dataset that contains all the fixes to CVEs. I believe it can help this project. Dataset, code and the paper are publicly available:
https://dl.acm.org/doi/abs/10.1145/3663533.3664036

Let me know if you have any questions

**Comment by pombredanne** [comment]
@JafarAkhondali https://dl.acm.org/doi/abs/10.1145/3663533.3664036 looks wonderful!

**Comment by pombredanne** [comment]
@JafarAkhondali 

- I assume all the code including the modified prospector is in there https://zenodo.org/records/11110595 ? is there a git repo with the code proper? 
- What is the license for the data at https://zenodo.org/records/11199120 ? I see "free" on https://dl.acm.org/do/10.5281/zenodo.11199120/full/ but "free" is not a license and I cannot integrate data without a license

**Comment by JafarAkhondali** [comment]
I'll add the code in https://github.com/JafarAkhondali/morefixes soon.
The license is one of the limitations, I didn't add a restricted license for the dataset, however the license of the extracted projects is various and different.

**Comment by pombredanne** [comment]
@JafarAkhondali Thanks...
You wrote:

>  The license is one of the limitations, I didn't add a restricted license for the dataset, however the license of the extracted projects is various and different.

There are two things: the license of individual patches which is that of the code they originally belong to, and separately, the license of the database collection of patches that you created: this is this second license that I am interested in and that I need. It can be a CC0-1.0, a CC-BY-40, a CC-BY-SA-4.0 or anything, but I need a license to integrate this in VulnerableCode.

**Comment by JafarAkhondali** [comment]
@pombredanne This is on the Zenodo link(https://zenodo.org/records/11199120) 
<img width="468" alt="image" src="https://github.com/user-attachments/assets/4b2f6af6-6e11-4b53-a521-e6a07c87e069">
Is it enough or you need the license to be in another place?

**Comment by pombredanne** [comment]
@JafarAkhondali I had missed that. So I reckon that the data collection is under CC-BY-4.0 ?

**Comment by JafarAkhondali** [comment]
Yes, but tbh, I'm not so familiar with licensee. If there is something that I can handle, I'm open to make the dataset as OPEN as possible for any usage. The only restriction is that some codes belongs to projects that I don't own.

**Comment by ziadhany** [comment]
According to my understanding , Let’s say we are finding the paths in the affected package’s code graph that lead to the vulnerable functions.

We are given a VCS URL along with the commit hash and diff of the fix commit.

Steps:
1. Clone the repository from the VCS URL.
2. Check out the Parent of the Fix Commit ( commit immediately before the fix )
3. Calculate all reachable functions:

```
atom -o app.atom -l python .
cdxgen -t python --deep -o bom.json .
atom reachables -o app.atom -s reachables.json -l python .
```

4. Parse the diff to identify the affected functions 
5. Match the corresponding affected functions to `reachables.json` in the code graph to find the vulnerable functions.


Or may be, can we calculate the reachables functions for both the fix commit and the parent of the fix commit, and consider the diff as representing the vulnerable functions.
